Astro alerts use case (#3335)
astronomer/docs: astro/use-cases/airflow-astro-alerting.md

When you combine Airflow notifications with Astro alerts, you can:

- Easily configure DAG failure or success alerts for common communication channels like Slack, PagerDuty, or email with Astro alerts.
- Leverage Airflow email notifications and callbacks for custom task-level alerting logic.
- Implement pipeline SLAs based on task duration or absolute time with Astro alerts.

This use case shows how to combine Airflow and Astro alerts for a couple of common scenarios. For general guidance on when to use which type of alert, see the [Explanation](#explanation) section.

## Feature overview

This use case depends on the following Astro and Airflow features to create different types of alerts for your pipelines:

- Configuring DAG SLAs and failure notifications using [Astro alerts](alerts.md).
- Implementing [Airflow callbacks](https://docs.astronomer.io/learn/error-notifications-in-airflow#airflow-callbacks) for custom task-level alerts. 

## Prerequisites

This use case assumes you have:

> - At least one [Astro Deployment](create-deployment). Your deployment must run Astro Runtime 7.1.0 or later and it must have OpenLineage enabled.

Reviewer: ```suggestion
- At least one [Astro Deployment](create-deployment.md). Your deployment must run Astro Runtime 7.1.0 or later and it must have OpenLineage enabled.
```

- An [Astro project](https://docs.astronomer.io/astro/cli/develop-project) with at least one DAG. Your DAG should have at least two tasks.

However, you can extend this use case to encompass any number of Astro Deployments and DAGs.

## Implementation

To implement this use case:

> 1. In the DAG in your Astro project, add an `on_success_callback` that uses the pre-built SlackNotifier to one task (not the DAG's `default_args`). See [Example pre-built notifier: Slack](https://docs.astronomer.io/learn/error-notifications-in-airflow#example-pre-built-notifier-slack).  

Reviewer 2: ```suggestion
1. Open the DAG in your Astro project and configure a pre-built SlackNotifier for one of its tasks as a task-level argument. See [Example pre-built notifier: Slack](https://docs.astronomer.io/learn/error-notifications-in-airflow#example-pre-built-notifier-slack) for sample code.  
```

> 2. Deploy your project to your Astro Deployment. See [Deploy code to Astro](https://docs.astronomer.io/astro/deploy-code).

Reviewer 2: ```suggestion
2. Deploy your project to your Astro Deployment. See [Deploy code to Astro](deploy-code.md).
```

Author: @jwitz I had this before but the build failed because it couldn't resolve the link - I'm assuming because it's in a different folder? Not sure if there's a way around that?

> 3. Add a connection to Slack in the Astro Cloud UI. See [Create Airflow connections in the Cloud UI](https://docs.astronomer.io/astro/create-and-link-connections). This connection will be used by your SlackNotifier. Make sure the connection ID matches what you used in your DAG code in Step 1.

Reviewer 2: ```suggestion
3. Add a connection to Slack in the Astro Cloud UI. See [Create Airflow connections in the Cloud UI](create-and-link-connections.md). This connection will be used by your SlackNotifier, so make sure the connection ID matches what you used in your DAG code in Step 1.
```

> 4. Configure an Astro DAG failure alert for your DAG using the communication channel of your choice. See [Set up Astro alerts](https://docs.astronomer.io/astro/alerts).

Reviewer 2: ```suggestion
4. In your Astro Deployment, configure an Astro **DAG failure** alert for your DAG using the communication channel of your choice. See [Set up Astro alerts](https://docs.astronomer.io/astro/alerts).
```

> 5. Configure an Astro absolute time alert for your DAG based on the amount of time you expect your DAG to complete in. Use the communication channel of your choice. See [Set up Astro alerts](https://docs.astronomer.io/astro/alerts).

Reviewer 2: ```suggestion
5. Configure an Astro **Absolute time** alert for your DAG based on the amount of time you expect your DAG to complete in. Use the communication channel of your choice. See [Set up Astro alerts](https://docs.astronomer.io/astro/alerts).
```

## Explanation

> Using a combination of Astro alerts and Airflow notifications allows your team to easily implement alerting logic for any scenario. You may need different types of alerts that are better suited towards one option or the other, as shown in the example in this use case:

Reviewer 2: ```suggestion
Using a combination of Astro alerts and Airflow notifications allows your team to implement alerting logic using the best tool for each scenario. You might need different types of alerts that are better suited towards one option or the other, as shown in the example in this use case:
```

- Task-level alerts with custom logic, for example if you want to be notified of success or failure only for a specific task, or want to run custom code if a task succeeds or fails, are straightforward to implement in Airflow. Task-level success or failure alerts or alerts with custom logic are not currently supported with Astro alerts.
- DAG-level success or failure alerts can be accomplished using either Airflow notifications or Astro alerts, but on Astro they are easy to configure and do not require any code changes. To implement these alerts with Airflow, you would need to update your DAG code and possibly complete more configuration for your communication channel (e.g. setting up an SMTP server for email alerts).

> - Absolute time alerts are straightforward to implement with Astro alerts, but are not currently supported in Airflow.

Reviewer 3: Maybe reword this as: "Absolute time alerts, or alerts where you don't want to wait for the task's state to change, should be done via the Astro alerts." Stylistic choice

Author: What do you mean by "wait for the task's state to change"? Isn't that the case with any SLA? Regardless, how about "Absolute time alerts should be implemented using Astro alerts. They are not currently supported in Airflow."

Want to be clear that Airflow doesn't have this mechanism today.

> - Absolute time alerts are straightforward to implement with Astro alerts, but are not currently supported in Airflow.

Author: ```suggestion
- Absolute time alerts, for when you need your DAG to complete by a certain time of day, should be implemented using Astro alerts. They are not currently supported in Airflow
```

> - Task-level alerts with custom logic, for example if you want to be notified of success or failure only for a specific task, or want to run custom code if a task succeeds or fails, are straightforward to implement in Airflow. Task-level success or failure alerts or alerts with custom logic are not currently supported with Astro alerts.
> - DAG-level success or failure alerts can be accomplished using either Airflow notifications or Astro alerts, but on Astro they are easy to configure and do not require any code changes. To implement these alerts with Airflow, you would need to update your DAG code and possibly complete more configuration for your communication channel (e.g. setting up an SMTP server for email alerts).
> - Absolute time alerts, for when you need your DAG to complete by a certain time of day, should be implemented using Astro alerts. They are not currently supported in Airflow

Reviewer 2: ```suggestion
- **Task-level alerts with custom logic** are recommended if you want be notified of success or failure only for a specific task, or you want to run custom code if a task succeeds or fails. Task-level success or failure alerts or alerts with custom logic are not currently supported with Astro alerts, but they're straightforward to implement in Airflow
- DAG-level success or failure alerts can be created using either Airflow notifications or Astro alerts, but on Astro they are easy to configure and don't require any code changes. To implement these alerts with Airflow, you need to update your DAG code and complete additional configuration for your communication channel (for example, setting up an SMTP server for email alerts).
- Absolute time alerts are recommended when you need your DAG to complete by a certain time of day. These types of alerts are available only on Astro.
```

> In some cases, whether you set up an alert on Astro or in Airflow will be dependent on how your DAGs were configured before you moved to Astro. For example, if all of your DAGs have email on failure notifications configured already, it will likely be easier for you to configure an SMTP server on Astro and rely on the existing notification rather than implementing a new Astro alert for each DAG. On the other hand, if you are developing new DAGs on Astro, it will likely be easier to implement Astro alerts which require less configuration for the communication channels.

Reviewer 2: ```suggestion
In some cases, whether you set up an alert on Astro or in Airflow will depend on how your DAGs were configured before you moved to Astro. For example, if all of your DAGs already have email-on-failure notifications configured, it will likely be easier for you to configure an SMTP server on Astro and rely on the existing notification than implement a new Astro alert for each DAG. On the other hand, if you are developing new DAGs on Astro, it will likely be easier to implement Astro alerts which require less configuration for the communication channels.
```

> :::note

Reviewer 2: ```suggestion
:::info
```

Programmatically creating Astro alerts, which may be helpful if you have a large number of DAGs, is coming soon.

:::

> In other cases, Astro alerts are clearly the better option. If you need to implement SLAs for your pipelines so you know as soon as a task is taking too long or a DAG hasn't completed on time, you should use the absolute time or task duration Astro alerts. There are a couple of reasons these are preferable to Airflow SLAs:

Reviewer 2: ```suggestion
In other cases, Astro alerts are always the better option. If you need to implement SLAs for your pipelines so you immediately know when a DAG task is taking too long to run, you should use the absolute time or task duration Astro alerts. There are a couple of reasons these are preferable to Airflow SLAs:
```

- The Astro alert for absolute time allows you to implement alerts based on a specific time of day your DAG should complete. This type of SLA is not available in Airflow.
- Astro alerts allow you to easily choose between multiple common communication channels. With Airflow SLAs, if you want your notification sent somewhere other than your email, you have to write your own custom callback logic. Additionally, if you have an SMTP service configured in your Airflow environment you will get email notifications for SLA misses; there is no way to turn this behavior off.

### Recommended alerts

Generally, Astronomer recommends choosing the following alerts for these common scenarios:

| Scenario                                   | Astro alert | Airflow email notification | Airflow callback | Airflow SLA | Airflow timeout |
|--------------------------------------------|-------------|-----------------------------|------------------|--------------|------------------|

> | Email on DAG success or retry               | X           | X                           |                  |              |                  |

Author: Because until we have a programmatic way of creating alerts, this one could truly go either way. I have a paragraph somewhere explaining this, but if you were moving on to Astro and had 100 DAGs with email alerts already set up, it's definitely easier to just set up SMTP on Astro than to create 100 different Astro alerts. If you have a smaller number of DAGs, or don't already have alerts configured, then Astro alerts are preferable. Once adding bulk alerts via API or something like that is available in the product, I'd be more comfortable suggesting Astro alerts explicitly for this case.
